sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColumns.java

```diff
@@ -70,21 +77,55 @@ public GeneratedColumns(InputFactory inputFactory,
         toValidate = validation == Validation.NONE ? Collections.emptyMap() : new HashMap<>();
         InputFactory.Context<CollectExpression<T, ?>> ctx = inputFactory.ctxForRefs(txnCtx, refResolver);
         toInject = new HashMap<>();
+
         for (Reference colWithDefault : allDefaultExpressionColumns) {
             if (!presentColumns.contains(colWithDefault)) {
                 toInject.put(colWithDefault, ctx.add(colWithDefault.defaultExpression()));
             }
         }
-        for (GeneratedReference generatedCol : allGeneratedColumns) {
-            if (presentColumns.contains(generatedCol)) {
-                if (validation == Validation.VALUE_MATCH) {
-                    toValidate.put(generatedCol, ctx.add(generatedCol.generatedExpression()));
+
+        maybeInjectNested = new HashMap<>();
+        Set<String> topLevelPresentColumns = presentColumns
+            .stream()
+            .map(Reference::column)
+            .filter(ColumnIdent::isTopLevel)
+            .map(ColumnIdent::name)
+            .collect(Collectors.toSet());
+
+        allGeneratedColumns.forEach(
+            r -> processGeneratedExpressionColumn(
+                r,
+                ctx.add(r.generatedExpression()),
+                presentColumns,
+                topLevelPresentColumns,
+                validation == Validation.VALUE_MATCH)
+        );
+        expressions = ctx.expressions();
+    }
+
+    private void processGeneratedExpressionColumn(Reference reference,
+                                                 Input<?> expressionInput,
+                                                 Collection<Reference> presentColumns,
+                                                 Set<String> topLevelPresentColumns,
+                                                 boolean doValidate) {
+        final ColumnIdent ident = reference.column();
+        final boolean isNestedColumn = !ident.isTopLevel();
+        if (isNestedColumn) {
```

Comment:
I think this is only true on update, can we add a test please?

Reply:
it can actually happen with a nested generated column, the added tests cover such cases.